content/blog/2020-05-22-rt-template/index.md

```diff
@@ -0,0 +1,365 @@
+
+[grad]: https://www.cs.cornell.edu/~asampson/blog/llvm.html
+[skeleton]: https://github.com/sampsyo/llvm-pass-skeleton
+[leech]: https://github.com/sa2257/llvm-runtime-pass
+
+## What this framework is targeting
+I'm defining some key levels to insert a runtime function replacements.
+- Instruction level
+- Subgraph level
+- Basic block level
+- Function level
+This framework will provide a pass for each of them and provide some rudimentary tools
+needed for a replacement at each level. Ideally, these tools should be generic and complete.
+However, the replacements are highly dependent on the program and the runtime function.
+Hence this is an experiment to evaluate how robust replacements at each level.
+
+## Replacing instructions
+Instructions provide a clearly defined interface for the runtime functions to call.
+Therefore, replacing an instruction is probably the easiest pass to implement, and also 
+the easiest to extend to a substantial benchmark suite. 
+
+```c
+    /* create the function call from runtime library */
+    LLVMContext& Ctx = F.getContext();
+    FunctionCallee rtFunc;
+    rtFunc = F.getParent()->getOrInsertFunction(
+        "ins_replace", 
+        Type::getDoubleTy(Ctx),
+        Type::getDoubleTy(Ctx), Type::getDoubleTy(Ctx), Type::getInt32Ty(Ctx)
+        );
+    // function call name, return type and argument types have to be known
+
+    for (auto& B : F) {
+        queue <Instruction*> workList; // collects replaced instructions within iterator
+        for (auto& I : B) {
+            /* Add runtime instruction selection here */
+            
+            /* End of selection algorithm */
+                
+            /* find the place to enter the runtime call */
+            IRBuilder<> builder(I);
+            builder.SetInsertPoint(&B, ++builder.GetInsertPoint());
+            
+            
+            /* insert runtime call, collect return value */
+            Constant *i32_config = ConstantInt::get(Type::getInt32Ty(Ctx), /* configuration */, true);
+            Value* args[] = {I.getOperand(0), I.getOperand(1), i32_config};
+            Value* ret =  builder.CreateCall(rtFunc, args);
+            
+            /* forward return to users */
+            for (auto& U : I.uses()) {
+                User* user = U.getUser();
+                user->setOperand(U.getOperandNo(), ret);
+            }
+            
+            /* mark instruction for deletion */
+            workList.push(&I);
+            I.dropAllReferences();
+        }
+        
+        /* delete replaced instructions */
+        while (!workList.empty()) {
+            workList.front()->eraseFromParent();
+            workList.pop();
+        }
+    }
+```
+Key limitations to note, runtime library has to be co-designed with the function declarations 
+baked into the llvm pass and of course, the instruction selection algorithm can get very complex
+depending on how interesting the task is.
```

Comment:
What do you mean by "co-designed"? Can you explain this in an independent way that doesn't require the reader to already understand your code?